AK/Error.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2023, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Error.h"
+#include "String.h"
+
+namespace AK {
+
+Error::Error(Error&& other)
+{
+    swap(this->m_syscall, other.m_syscall);
+    swap(this->m_code, other.m_code);
+#if !defined(KERNEL)
+    if (other.m_string_literal.has<StringView>())
+        this->m_string_literal = move(other.m_string_literal.get<StringView>());
+    else if (other.m_string_literal.get<String const*>() != nullptr) {
+        this->m_string_literal = move(other.m_string_literal.get<String const*>());
+        // Make sure the other Error doesn't destruct the string!
+        other.m_string_literal.get<String const*>() = nullptr;
+    }
+#else
+    swap(this->m_string_literal, other.m_string_literal);
+#endif
+}
+
+Error::Error(Error const& other)
+    : m_code(other.m_code)
+    , m_syscall(other.m_syscall)
+{
+
+#if !defined(KERNEL)
+    if (other.m_string_literal.has<StringView>())
+        this->m_string_literal = other.m_string_literal.get<StringView>();
+    else if (other.m_string_literal.get<String const*>() != nullptr) {
+        this->m_string_literal = new (nothrow) String { *other.m_string_literal.get<String const*>() };
+    }
+#else
+    this->m_string_literal = other.m_string_literal;
+#endif
+}
+
+Error::Error(Error& other)
+    : Error(static_cast<Error const&>(other))
+{
+}
+
+Error::~Error()
+{
+#if !defined(KERNEL)
+    if (m_string_literal.has<String const*>() && m_string_literal.get<String const*>() != nullptr)
+        delete m_string_literal.get<String const*>();
+#endif
+}
+
+#if !defined(KERNEL)
+Error::Error(String const* owned_string_literal)
+    : m_string_literal(owned_string_literal)
+{
+    VERIFY(m_string_literal.has<String const*>());
+}
+
+Error::Error(String const* owned_string_literal, int code)
+    : m_string_literal(owned_string_literal)
+    , m_code(code)
+{
+    VERIFY(m_string_literal.has<String const*>());
+}
+
+Error Error::from_string(ErrorOr<String>&& string_or_error, Optional<int> code)
+{
+    // We intentionally do not return the String creation error here, since that is probably unrelated to the actual error.
+    if (string_or_error.is_error()) {
+        if (code.has_value())
+            return { code.value() };
+        return { ""sv };
```

Comment:
Not sure this is a good fallback, we get no information here and will have no idea what the error is / where it came from. Not sure what to do though, short of adding a `SourceLocation` to `Error`.

Reply:
There's no general-purpose errno, so we have no good option here. I don't want to propagate the string error since that's most likely unrelated.

Reply:
I will enforce a StringView fallback.